frame/collective/src/lib.rs

```diff
@@ -182,72 +306,200 @@ decl_module! {
 		///
-		/// - `new_members`: The new member list. Be nice to the chain and
-		//	provide it sorted.
+		/// - `new_members`: The new member list. Be nice to the chain and provide it sorted.
 		/// - `prime`: The prime member whose vote sets the default.
 		///
 		/// Requires root origin.
-		#[weight = (100_000_000, DispatchClass::Operational)]
-		fn set_members(origin, new_members: Vec<T::AccountId>, prime: Option<T::AccountId>) {
+		///
+		/// # <weight>
+		/// ## Weight
+		/// - `O(MP + N)` where:
+		///   - `M` old-members-count (code- and governance-bounded)
+		///   - `N` new-members-count (code- and governance-bounded)
+		///   - `P` proposals-count (code-bounded)
+		/// - DB:
+		///   - 1 storage mutation (codec `O(M)` read, `O(N)` write) for reading and writing the members
+		///   - 1 storage read (codec `O(P)`) for reading the proposals
+		///   - `P` storage mutations (codec `O(M)`) for updating the votes for each proposal
+		///   - 1 storage write (codec `O(1)`) for deleting the old `prime` and setting the new one
+		/// - Benchmark: 0 + M * 20.47 + N * 0.109 + P * 26.29 µs (min squares analysis)
+		/// # </weight>
+		#[weight = (
+			weight_for::set_members(
+				T::DbWeight::get(),
+				T::MaxMembers::get(), // M
+				new_members.len() as Weight, // N
+				T::MaxProposals::get(), // P
+			),
+			DispatchClass::Operational
+		)]
+		fn set_members(
+			origin, new_members: Vec<T::AccountId>, prime: Option<T::AccountId>
+		) -> DispatchResultWithPostInfo {
 			ensure_root(origin)?;
+			ensure!(new_members.len() <= T::MaxMembers::get() as usize, Error::<T, I>::TooManyMembers);
+
+			let old = Members::<T, I>::get();
 			let mut new_members = new_members;
 			new_members.sort();
-			let old = Members::<T, I>::get();
 			<Self as ChangeMembers<T::AccountId>>::set_members_sorted(&new_members[..], &old);
 			Prime::<T, I>::set(prime);
+
+			Ok(Some(weight_for::set_members(
+				T::DbWeight::get(),
+				old.len() as Weight, // M
+				new_members.len() as Weight, // N
+				T::MaxProposals::get(), // P
+			)).into())
 		}
 
 		/// Dispatch a proposal from a member using the `Member` origin.
 		///
 		/// Origin must be a member of the collective.
-		#[weight = (100_000_000, DispatchClass::Operational)]
-		fn execute(origin, proposal: Box<<T as Trait<I>>::Proposal>) {
+		///
+		/// # <weight>
+		/// ## Weight
+		/// - `O(M + P)` where `M` members-count (code-bounded) and `P` complexity of dispatching `proposal`
+		/// - DB: 1 read (codec `O(M)`) + DB access of `proposal`
+		/// - 1 event
+		/// - Benchmark: 22.62 + M * 0.115 + B * 0.003 µs (min squares analysis)
+		/// # </weight>
+		#[weight = (
+			weight_for::execute(T::DbWeight::get(), T::MaxMembers::get(), proposal.get_dispatch_info().weight),
+			DispatchClass::Operational
+		)]
+		fn execute(origin, proposal: Box<<T as Trait<I>>::Proposal>) -> DispatchResultWithPostInfo {
 			let who = ensure_signed(origin)?;
-			ensure!(Self::is_member(&who), Error::<T, I>::NotMember);
```

Comment:
consider changing the is `is_member` to input the member list, so you can keep the checks consistent here. I am thinking worst case scenario `is_member` changes and this check does not.

Reply:
`is_member` is actually not used anymore. It is a public function, though, so I don't know whether it's still used somewhere.

Reply:
As far as I see is_member is not used in the codebase though it seems maybe usefull API.

We can do:
* remove it
* add is_member_inner with an additional field
* just let is as is.

I think third option is fine is_member should be considered public API not an internal abstraction to check member so if one change this they should be careful.
you can add an internal comment stating this though.

Reply:
Added a comment.